### PR TITLE
Fix weights not properly initialized due to shape mismatch

### PR DESCRIPTION
**Pull request by ydshieh**
Currently, if there is some weight shape mismatched between the model and the checkpoint, and if `ignore_mismatched_sizes=True`, that/those weight(s) won't get initialized by the model's `_init_weights` method, and could get crazy values like `1e37`.

This could make the training gets nan loss value from the beginning and won't have any progress.

One example is by running `src/transformers/modeling_utils.py` (add `ignore_mismatched_sizes=True`).

We usually set `ignore_mismatched_sizes=True` when we want to perform classification tasks using an existing model but to another task having different number of targets.

This PR aims to fix this issue.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -539,15 +539,24 @@ def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
...
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if len(set(module.state_dict().keys()) - set(loaded_keys)) == 0:
+                module._is_hf_initialized = loaded
+        else:
+            not_loaded_keys = [
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if len(set(module.state_dict().keys()).intersection(not_loaded_keys)) > 0:
+                module._is_hf_initialized = False
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:559]: We do this conversion in several places. So a (inner) function to serve them.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -3955,14 +3964,22 @@ def _fix_key(key):
...
+                # The model key starts with `prefix` but `checkpoint_key` doesn't so we add it.
+                model_key = f"{prefix}.{key}"
             elif add_prefix_to_model:
-                _loaded_keys = [k[len(prefix) + 1 :] for k in loaded_keys]
-            else:
-                _loaded_keys = loaded_keys
+                # The model key doesn't start with `prefix` but `checkpoint_key` does so we remove it.
+                model_key = key[len(prefix) + 1 :]
+
+            return model_key
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3976]: We do this conversion in several places. So a (inner) function to serve them.

File: src/transformers/modeling_utils.py (hunk 1 above)
Comment (ydshieh) [inline src/transformers/modeling_utils.py:559]: We want to be able to set some module(s) as `not initialized` in the case we have some checkpoint/model weight(s) having different shape(s) - that **we can only find later**.

In this case, currently on `main`, we have:

- this block marks them as initialized (but not really initialize the weight)
- later, when we want to set the model weight by the checkpoint weight, it is skipped as the shape doesn't match (if `ignore_mismatched_sizes=True`)
  - so that/those weights are not initialized by the model's `_init_weights` and could get crazy values like `1e37`, and cause training issue.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -3955,14 +3964,22 @@ def _fix_key(key):
...
+
+        # retrieve unintialized modules and initialize before maybe overriding that with the pretrained weights.
+        if _fast_init:
+            _loaded_keys = [
+                checkpoint_key_to_model_key(k, remove_prefix_from_model, add_prefix_to_model) for k in loaded_keys
+            ]
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3982]: just simplified by the new method `checkpoint_key_to_model_key` above

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -4004,13 +4021,9 @@ def _find_mismatched_keys(
...
-                    model_key = checkpoint_key
-                    if remove_prefix_from_model:
-                        # The model key starts with `prefix` but `checkpoint_key` doesn't so we add it.
-                        model_key = f"{prefix}.{checkpoint_key}"
-                    elif add_prefix_to_model:
-                        # The model key doesn't start with `prefix` but `checkpoint_key` does so we remove it.
-                        model_key = ".".join(checkpoint_key.split(".")[1:])
+                    model_key = checkpoint_key_to_model_key(
+                        checkpoint_key, remove_prefix_from_model, add_prefix_to_model
+                    )
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:4026]: same

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -4157,6 +4170,15 @@ def _find_mismatched_keys(
...
                 shutil.rmtree(state_dict_folder)
 
+        if _fast_init:
+            mismatched_model_keys = [
+                checkpoint_key_to_model_key(x[0], remove_prefix_from_model, add_prefix_to_model)
+                for x in mismatched_keys
+            ]
+            set_initialized_submodules(model, mismatched_model_keys, loaded=False)
+            # This will only initialize submodules that are re-marked as `not loaded` above due to mismatched
+            model.apply(model._initialize_weights)
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:4180]: Look at the `mismatched_model_keys` and  initialize the weights for them

**Review by ydshieh — COMMENTED** [review]
Some comments

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -3955,14 +3964,22 @@ def _fix_key(key):
                             model, key, "cpu", torch.empty(*param.size(), dtype=target_dtype)
                         )
 
-        # retrieve unintialized modules and initialize before maybe overriding that with the pretrained weights.
-        if _fast_init:
+        def checkpoint_key_to_model_key(key, remove_prefix_from_model, add_prefix_to_model):
+            model_key = _fix_key(key)
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3968]: In #16920, `original_loaded_keys` is used when calling `_find_mismatched_keys`. Before that PR, it was `loaded_keys` being used (the one had went through `_fix_key`). 

That PR forgot to recompute the `model_key` via `_fix_key` inside `_find_mismatched_keys`.

Here, we fix this issue too.

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (ydshieh) [inline src/transformers/modeling_utils.py:3976]: We do this conversion in several places. So a (inner) function to serve them.

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -539,15 +539,24 @@ def load_state_dict(checkpoint_file: Union[str, os.PathLike]):
...
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if len(set(module.state_dict().keys()) - set(loaded_keys)) == 0:
+                module._is_hf_initialized = loaded
+        else:
+            not_loaded_keys = [
+                k.replace(f"{module_name}.", "") for k in state_dict_keys if k.startswith(f"{module_name}.")
+            ]
+            if set(module.state_dict().keys()) == set(not_loaded_keys):
+                module._is_hf_initialized = False
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:559]: We want to be able to set some module(s) as not initialized in the case we have some checkpoint/model weight(s) having different shape(s) - that we can only find later.

In this case, currently on main, we have:

- this block marks them as initialized (but not really initialize the weight)
- later, when we want to set the model weight by the checkpoint weight, it is skipped as the shape doesn't match (if ignore_mismatched_sizes=True)
  - so that/those weights are not initialized by the model's _init_weights and could get crazy values like 1e37, and cause training issue.

Comment (ydshieh) [reply src/transformers/modeling_utils.py:559]: The condition `set(module.state_dict().keys()) == set(not_loaded_keys)` is to make sure we only initialize that module when all its weights are not yet initialized (due to shape mismatched).

If some weights of a module are loaded from checkpoint but some of its weights are not loaded due to shape mismatch, we don't really have a way to proceed: as a model's `_init_weight` method can only operation on a module, not its individual weights.

**Comment by ydshieh** [comment]
As this touches the core file, request 2 reviews 🙏

**Comment by ydshieh** [comment]
Guess I have to add some tests for this case. Change to draft for now
